### ISSUE TITLE
End to End testing and benchmarks

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/ipfs-cluster#12]
Need to automatically bring up a real cluster, in a real cloud/hw, with real IPFS to and perform a number of standard cluster workloads.

Any measures extracted from these tests can be used as future reference regarding the performance of the Cluster.

**Comment by VictorBjelkholm** [comment]
Would love to do this together with in relation to kubernetes-ipfs, think it would make a lot of sense to provide a deployment that people can use easily, and that we can put under E2E tests.

**Comment by hsanjuan** [comment]
@VictorBjelkholm yes, we need to have a conversation about this. kubernetes-ipfs looks promising, but right now I would miss the possibility of being able to deploy a testing cluster to X provider (my raspberry pi cluster, aws, DO, Kubernetes) without much hassle... it seemed kubernetes-ipfs mostly addressed local deployment.

I was impressed with the dashboard and stats though, and I am not familiar with Kubernetes so I need to take a closer look to see how it can serve me. It's definitely worth to align efforts and make things re-usable across projects.

**Comment by jbenet** [comment]
kubernetes should be able to run on rpis, aws, DO fine. the kubernetes server does not work on non-linux though, i think.

**Comment by hsanjuan** [comment]
Note to self and everyone: before the https://github.com/ipfs/pm/issues/354 sprint, we should have a list of cluster tests which would be nice to have.

**Comment by hsanjuan** [comment]
Ok I start the list:

X ranges from 3 to ...
Y ranges from 1 to ...

* Start a X-peer cluster. Check that everyone becomes ready. Check that all peers know all other peers.
* Start a X-peer cluster with replication factor -1. Make Y pins to random nodes with random content. Check that everyone knows about all pins. Check that they are in PINNED state. Benchmark.
* Start a X-peer cluster with replication factor 3. Make Y pins to random nodes with random content. Check that everyone knows about pins. Check that they are PINNED in three peers and the rest is REMOTE.
* Start a X-peer cluster. Make Y pins. Block network for 1/3 of peers. Make Y unpins to random nodes in the 2/3 side. Check that nothing is pinned in 2/3 of cluster. Unblock network. Check that nothing is pinned anywhere.
* Start a X-peer cluster. Block 2/3rds of peers. Make Y pins to the remaining 1/3. Ensure all error.
* Start a X-peer cluster with replication factor 3. Make Y pins to random nodes. Kill some IPFS daemons. Check that pins become repinned.
* Start a X-peer cluster with replication factor 3. Make Y pins to random nodes. Block network for some cluster nodes. Ensure that pins get correctly replicated to other peers.
* Start a X-peer cluster. Pin Y large files. Ensure everything becomes PINNED (according to replication factor) eventually (there's a 15-minute-become-pin-error timeout in the PinTracker which might need adjust).
* Start a X-peer cluster. Make Y pins to random nodes with random content. Unpin everything in random nodes all at once. Check that nothing is pinned.
* Start a X-peer cluster. Make Y pins.  Start Z separate peers. Run  PeerAdd() operations for those peers in the same cluster peer. Check that all peers know about all peers and about the pins. Run PeerRm() operations in the same peer. Make sure everyone knows about the departures.
* Same as above except running PeerAdd and PeerRm on random peers.
* Start a X-peer cluster. Make Y pins while randomly killing and restarting peers (after a random interval). Keep doing this for a while. Bring everyone up. Ensure that all the pins that suceeded are known by everyone and correctly replicated.

---

* Pick some of the above tests and re-run them in a network with very big latency/random packet drops/very low bandwidth.

**Comment by hsanjuan** [comment]
@jbenet I have a piece of paper saying you wanted to be reminded to give some thought to this. Do you want to add some?

**Comment by hsanjuan** [comment]
@ZenGround0 I think we can tick a few more tests from my list :)
